add DEPLOY_PATH support

scytacki opened the pull request:
Building an index-top.html that has a publicPath of DEPLOY_PATH, is required to support releasing with the s3 deploy github action.

This PR sets up that support and fixes places where assets are loaded relative to the index.html so they will instead be loaded relative to the public path.

This PR combined with this other PR:
https://github.com/concord-consortium/collaborative-learning/pull/1455

Should give CLUE the support for the new deployment approach.

- [x] with this change stamps in the drawing tool cause new firebase images to be created each time the drawing tool is loaded. This is because the stamp ends up with a absolute URL. And the ImageMap automatically uploads absolute URLs to firebase. However the drawing tools image objects do not store the uploaded url `ccimg://fbrtdb...`, so each time they request a new entry from the image map the content is uploaded again. We need a solution for this, perhaps identify asset urls better in the ImageMap. Or update the drawing tool so it replaces the absolute URL with the firebase URL.
- [x] check tile and document export of tiles with images. This code seems to make assumptions about images with filenames. If the url is absolute perhaps it will break something here.

kswenson commented on webpack.config.js:
```diff
@@ -217,11 +223,23 @@ module.exports = (env, argv) => {
         }
       }),
+      ...(DEPLOY_PATH ? [new HtmlWebpackPlugin({
+        filename: 'index-top.html',
+        template: 'src/index.html',
+        favicon: 'src/public/favicon.ico',
+        publicPath: DEPLOY_PATH,
+        templateParameters: {
+          rollbarSnippet,
+          ...packageJson.config
+        }
```
Seems like you could dry up the code a bit with something like:
```
      const baseHtmlPluginConfig = {
        template: 'src/index.html',
        favicon: 'src/public/favicon.ico',
        templateParameters: {
          rollbarSnippet,
          ...packageJson.config
        }
      };
```
and then...
```
      new HtmlWebpackPlugin({
        filename: 'index.html',
        publicPath: '.',
        ...baseHtmlPluginConfig
      }),
      ...(DEPLOY_PATH ? [new HtmlWebpackPlugin({
        filename: 'index-top.html',
        publicPath: DEPLOY_PATH,
        ...baseHtmlPluginConfig
      }
```

scytacki replied:
👍 I've made that change.

kswenson reviewed (approved):
👍 Looks good! Just one suggestion to DRY up the code a bit if you're so inclined.

scytacki commented:
To address the issue with the stamps, updated all the places that render stamps to use the ImageMap and I removed the pre process snapshot in the stamp object. So now the `stamp.url` will stay relative. And the urls of the ImageObject's in the drawing tool will have these same relative URLs. 

This also changed how legacy URLs are handled in the drawing tool:
Legacy URLs in drawing tool saved state (`assets/...`) will not be converted. The URLs are still handled, the saved state is just not updated. This approach seems best to move the legacy conversion all into a single place.

The approach of relative URLs means that if a stamp is removed from a future version of the curriculum the stamp will be broken in documents based on the old curriculum.  This is not new behavior it is is how it worked before.

I checked the export of document JSON and it seems to be working fine even when javascript and curriculum is in a different place than the index.html. What it seems to do is assume any image in the document is located in an images folder for the unit path. Despite my worries, it seems this behavior is not affected by changes in the PR.

scytacki commented:
@tejal-shah I added you as a reviewer so you could see what kinds of code and build system changes are needed to support this new approach to doing releases.  I think CLUE will be our most complicated project to migrate to the new s3 deployment since it allows user to upload images and has json content with relative references to images.  

The AP has json content with images but they aren't relative references. Question Interactives allows student to upload attachments but again these are handled as absolute URLs.  And I think customized stamps can be made in the Question Interactives drawing interactives but these are also absolute URLs.

There are some projects that download collections of resources from the project build, those might have some of these complications. I think some of the Geo projects do this.

kswenson reviewed (approved):
👍 Looks good!

tejal-shah commented:
Thanks @scytacki for including me! I'm wondering if we'll need to make similar changes for codap v3 as well where, not right now, but in the future, there will be .codap, .csv and image file downloads.

scytacki commented:
@tejal-shah, yes perhaps if there are some standard CODAP documents that are included in the CODAP repository and downloaded from the same place as the CODAP javascript files then they will need a similar treatment.  I'd guess CODAP doesn't use something like the ImageMap that CLUE has, so I'd guess these changes will be more simple.

kswenson commented:
>I'm wondering if we'll need to make similar changes for codap v3 as well where, not right now, but in the future, there will be .codap, .csv and image file downloads.

I don't think so. The issue that CLUE has is that it's storing references to curriculum content and to user-uploaded images in user content. When CODAP imports a document it just reads the data without storing any internal links to assets/resources, which is where the tricky issues come up.
